wal/wal.go

```diff
@@ -0,0 +1,825 @@
+// If an errore occurs during read, the repair procedure must be called
+// before it's safe to do further writes.
+//
+// Segments are written to in pages of 32KB, with records possibly split
+// across page boundaries.
+// Records are never split across segments to allow full segments to be
+// safely truncated. It also ensures that torn writes never corrupt records
+// beyond the most recent segment.
+type WAL struct {
+	dir         string
+	logger      log.Logger
+	segmentSize int
+	mtx         sync.RWMutex
+	segment     *Segment // active segment
+	donePages   int      // pages written to the segment
+	page        *page    // active page
+	stopc       chan chan struct{}
+	actorc      chan func()
+
+	fsyncDuration   prometheus.Summary
+	pageFlushes     prometheus.Counter
+	pageCompletions prometheus.Counter
+}
+
+// New returns a new WAL over the given directory.
+func New(logger log.Logger, reg prometheus.Registerer, dir string) (*WAL, error) {
+	return NewSize(logger, reg, dir, defaultSegmentSize)
+}
+
+// NewSize returns a new WAL over the given directory.
+// New segments are created with the specified size.
+func NewSize(logger log.Logger, reg prometheus.Registerer, dir string, segmentSize int) (*WAL, error) {
+	if segmentSize%pageSize != 0 {
+		return nil, errors.New("invalid segment size")
+	}
+	if err := os.MkdirAll(dir, 0777); err != nil {
+		return nil, errors.Wrap(err, "create dir")
+	}
+	if logger == nil {
+		logger = log.NewNopLogger()
+	}
+	w := &WAL{
+		dir:         dir,
+		logger:      logger,
+		segmentSize: segmentSize,
+		page:        &page{},
+		actorc:      make(chan func(), 100),
+		stopc:       make(chan chan struct{}),
+	}
+	w.fsyncDuration = prometheus.NewSummary(prometheus.SummaryOpts{
+		Name: "prometheus_tsdb_wal_fsync_duration_seconds",
+		Help: "Duration of WAL fsync.",
+	})
+	w.pageFlushes = prometheus.NewCounter(prometheus.CounterOpts{
+		Name: "prometheus_tsdb_wal_page_flushes_total",
+		Help: "Total number of page flushes.",
+	})
+	w.pageCompletions = prometheus.NewCounter(prometheus.CounterOpts{
+		Name: "prometheus_tsdb_wal_completed_pages_total",
+		Help: "Total number of completed pages.",
+	})
+	if reg != nil {
+		reg.MustRegister(w.fsyncDuration, w.pageFlushes, w.pageCompletions)
+	}
+
+	_, j, err := w.Segments()
+	if err != nil {
+		return nil, errors.Wrap(err, "get segment range")
+	}
+	// Fresh dir, no segments yet.
+	if j == -1 {
+		if w.segment, err = CreateSegment(w.dir, 0); err != nil {
+			return nil, err
+		}
+	} else {
+		if w.segment, err = OpenWriteSegment(w.dir, j); err != nil {
+			return nil, err
+		}
+		// Correctly initialize donePages.
+		stat, err := w.segment.Stat()
+		if err != nil {
+			return nil, err
+		}
+		w.donePages = int(stat.Size() / pageSize)
+	}
+	go w.run()
+
+	return w, nil
+}
+
+// Dir returns the directory of the WAL.
+func (w *WAL) Dir() string {
+	return w.dir
+}
+
+func (w *WAL) run() {
+	for {
+		// Processing all pending functions has precedence over shutdown.
```

Comment:
What happens if we shutdown with pending functions? The current code seems to allow for that